scheduler/scheduler.go

```diff
@@ -0,0 +1,128 @@
+package scheduler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/redis/go-redis/v9"
+)
+
+type StatusCode int
+
+const (
+	Creating StatusCode = iota
+	Ready
+	Running
+	Unknown
+)
+
+type FunctionLocation struct {
+	Address 	string		`redis:"address"`
+	Port    	uint16		`redis:"port"`
+	Status 		StatusCode 	`redis:"status"`
+	LastUsed	string		`redis:"lastUsed"`
+}
+
+
+type Scheduler struct {
+	Redis *redis.Client
+	Context *context.Context
+	Lambdo *LambdoService
+}
+
+// Uses SCAN to look for the first instance marked as ready (https://redis.io/docs/latest/commands/scan/)
+func (s *Scheduler) LookForReadyInstance(functionId uuid.UUID, cursor uint64) (id string, returnedCursor uint64, err error) {
+	locationMatch := fmt.Sprintf(functionId.String(), ":*")
+
+	keys, returnedCursor, err := s.Redis.Scan(*s.Context, cursor, locationMatch, 10).Result()
+
+	if err != nil {
+		return
+	}
+
+	for _, id := range keys {
+		code, err := s.Redis.HGet(*s.Context, id, "status").Int();
+
+		if err != nil && StatusCode(code) == Ready {
+			// we found a ready instance so we return it
+			return id, 0, nil
+		}
+	}
+
+	if returnedCursor != 0 {
+		// the current sweep did not find anything, we try again 
+		return s.LookForReadyInstance(functionId, returnedCursor)
```

Comment:
Is there a reason why this is recursive?

Reply:
Basically in Redis, SCAN allows you to get keys matching a certain pattern just like KEYS would. 

While KEYS locks the DB and scans everything in one shot, SCAN gets x keys and returns a cursor, you can then use the returned cursor to your next SCAN call to continue scanning until the returned cursor is 0.

This is why it is recursive, I can refactor it into interative if you want.

Reply:
Okay I understand
You don't need to put that in an iterative way but if could just add some comments in your code to say what you just explained it would be really nice